### PR TITLE
WaterLily + AutoDiff workflow

### PR DESCRIPTION
**Pull request by weymouth**
The goal of this PR is to enable auto-diff with WaterLily, ala issue #52. This PR is IN-PROGRESS!

A minor issue is that we have certain parts of the code which declare types which should not. I've removed some of these in the first commit.

The second and far more serious issue is that of the field arrays within `Flow` and `Poisson`. These are pre-allocated, and this is incompatible with ForwardDiff, which needs to be able to allocate Dual arrays for variables "downstream" of the derivative being taken. I've included an example "test.jl" file which fails with ForwardDiff when the simulation is pre-allocated as normal OR when pre-allocated with a Dual type.

The last issue is KernelAbstractions. It's possible that this will not be an issue once the types are sorted out, but the error messages are so bad that it's impossible to debug. I've changed `@loop` to just use `@SIMD` for the time being.

### CODE REVIEW

**Comment by b-fg** [comment]
I can reproduce the error too. I am looking into it.

**Comment by b-fg** [comment]
Nice, works as expected!

**Comment by b-fg** [comment]
KA with CPU backend is working. However, trying to use KA with CuArray is failing even if using: `θ = Ref(π/36) |> cudaconvert`, which returns a `CuRefValue`. And the error is arising at line `lift_hist = [step_force!(sim) for _ ∈ 1:20]`:
```
LoadError: MethodError: no method matching setindex!(::CUDA.CuRefValue{Float64}, ::Float64)

Closest candidates are:
  setindex!(::Ref, ::Any, ::CartesianIndex{0})
   @ Base multidimensional.jl:1933
```
So not very useful...

**Comment by weymouth** [comment]
That `Ref` problem on GPUs doesn't have anything to do with WaterLily in particular. We can play around with really small MWEs and see what works. Maybe post something on Slack. 

The second problem sounds like something we need to fix in the code base. Does that bit work on the CPU as is?

**Comment by b-fg** [comment]
Interestingly, with the new generalization of `@loop` to use `@simd` or KA based on `Threads.nthreads()`, I get a speedup when using a single thread compared to the bare `@simd` implementation, even though there are many more allocations. 
A)
```julia
macro loop(args...)
    ex,_,itr = args
    _,I,R = itr.args
    return quote
        @simd for $I ∈ $R
            @fastmath @inbounds $ex
        end
    end |> esc
end
```
B)
```julia
macro loop(args...)
    ex,_,itr = args
    _,I,R = itr.args; sym = []
    grab!(sym,ex)     # get arguments and replace composites in `ex`
    setdiff!(sym,[I]) # don't want to pass I as an argument
    @gensym(kern, kern_) # generate unique kernel function names for serial and KA execution
    return quote
        function $kern($(rep.(sym)...),::Val{1})
            @simd for $I ∈ $R
                @fastmath @inbounds $ex
            end
        end
        @kernel function $kern_($(rep.(sym)...),@Const(I0)) # replace composite arguments
            $I = @index(Global,Cartesian)
            $I += I0
            @fastmath @inbounds $ex
        end
        function $kern($(rep.(sym)...),_)
            $kern_(get_backend($(sym[1])),64)($(sym...),$R[1]-oneunit($R[1]),ndrange=size($R))
        end
        $kern($(sym...),Val{Threads.nthreads()}()) # dispatch to SIMD for -t 1, or KA otherwise
    end |> esc
end
```
resulting in the following benchmarks of `@btime sim_step!($sim)`
```
A) 776.913 μs (180 allocations: 250.23 KiB)
B) 656.911 μs (1376 allocations: 269.77 KiB)
```

**Comment by b-fg** [comment]
The simple SIMD version of `@loop`
```julia
macro loop(args...)
    ex,_,itr = args
    _,I,R = itr.args
    return quote
        @simd for $I ∈ $R
            @fastmath @inbounds $ex
        end
    end |> esc
end
```
Still shows 62 allocations for me. How did you write it to get 0 allocations @weymouth ?

**Comment by b-fg** [comment]
Interestingly enough, the simple `@loop` version produces 0 allocations, but is slower. 
Simple `@loop`
```
576.174 μs (0 allocations: 0 bytes)
```
Current "functional" `@loop`
```
434.004 μs (486 allocations: 7.84 KiB)
```

**Comment by weymouth** [comment]
@b-fg , What is the test case you are using for timing in those comments above?

**Comment by b-fg** [comment]
This one
```julia
using WaterLily, StaticArrays, BenchmarkTools

function make_sim(θ;L=32,U=1,Re=100,mem=Array)
    function map(x,t)
        s,c = sincos(θ)
        SA[c -s; s c]*(x-SA[L,L])
    end
    function sdf(ξ,t)
        p = ξ-SA[0,clamp(ξ[1],-L/2,L/2)]
        √(p'*p)-2
    end
    Simulation((2L,2L),(U,0),L,ν=U*L/Re,body=AutoBody(sdf,map),T=Float32,mem=mem)
end

sim = make_sim(π/36; Array)
sim_step!(sim)
@btime mom_step!($sim.flow, $sim.pois)
```
But idk what's happening that even with the simple `@loop` version I now get allocations on `mom_step!` using the same version (commit) I used before (!). Not sure what's going on.

**Comment by weymouth** [comment]
I can replicate that on my Windows PC at home. I'm getting 

`796.000 μs (0 allocations: 0 bytes)`

vs 

`626.200 μs (413 allocations: 6.70 KiB)`

I also checked the impact of array size using `sim = make_sim(0f0,L=256)` which gives

`31.799 ms (0 allocations: 0 bytes)`
vs

`27.886 ms (264 allocations: 4.38 KiB)`

So the amount of allocations _does not_ grow with array size and is consistently a little faster. Certainly not an issue.

**Comment by b-fg** [comment]
Are tests passing locally for `CuArrays` too? What else is needed for this PR?

File: src/Flow.jl [hunk 1]
```diff
@@ -61,21 +61,27 @@ upperBoundary!(r,u,Φ,ν,i,j,N,::Val{true}) = @loop r[I-δ(j,I),i] -= Φ[CIj(j,I
...
 end
-accelerate!(r,t,::Nothing,::Tuple) = nothing
+accelerate!(r,dt,::Nothing,::Tuple) = nothing
+"""
+    BCTuple(U,dt,N)
+
+Return BC tuple `U(i∈1:N, t=sum(dt))`.
+"""
+BCTuple(f::Function,dt,N,t=sum(dt))=ntuple(i->f(i,t),N)
+BCTuple(f::Tuple,dt,N)=f
```
Comment (b-fg) [inline src/Flow.jl:84]: These two definitions fo `BCTuple` are almost the same at lines 168-169 and the code does not compile for me.

**Comment by weymouth** [comment]
How can the code coverage be worse? I didn't change the number of lines or functions? I _reduced_ the number of arguments in two functions!!?

**Comment by b-fg** [comment]
It's also complaining about src/util.jl#L105, which you did not touch at all... Nvm, I think we are good.

**Comment by b-fg** [comment]
Doing some review, shouldn't the `T(0)` be `zero(T)` in
https://github.com/weymouth/WaterLily.jl/blob/b97d700dea197c2683f70d68753edaa591329bbf/src/WaterLily.jl#L71

File: src/util.jl [hunk 2]
```diff
@@ -228,8 +226,8 @@ function interp(x::SVector{D,T}, arr::AbstractArray{T,D}) where {D,T}
     end
     return s
 end
-function interp(x::SVector{D,T}, varr::AbstractArray{T}) where {D,T}
+function interp(x::SVector{D}, varr::AbstractArray) where {D}
     # Shift to align with each staggered grid component and interpolate
-    @inline shift(i) = SVector{D,T}(ifelse(i==j,0.5,0.) for j in 1:D)
-    return SVector{D,T}(interp(x+shift(i),@view(varr[..,i])) for i in 1:D)
+    @inline shift(i) = SVector{D}(ifelse(i==j,0.5f0,0f0) for j in 1:D)
```
Comment (b-fg) [inline src/util.jl:231]: Does this need to be Float32?

**Comment by b-fg** [comment]
Went over it once again, all looks good except for the 2 minor details above!

File: src/util.jl (hunk 2 above)
Comment (weymouth) [reply src/util.jl:231]: It doesn't seem to need to be, no.

**Comment by vchuravy** [comment]
@michel2323 has been working on adding Enzyme support to KernelAbstractions which together with Checkpointing.jl adds reverse mode capabilities.

We just gave a tutorial on using Enzyme for codes like this: https://dj4earth.github.io/MPE24 and in particular
https://dj4earth.github.io/MPE24/Burgers_tutorial/burgers_tutorial.html

We are still working through some minutia for full GPU support, but we are getting close.

**Comment by b-fg** [comment]
Thanks @vchuravy! That's very useful indeed. Currently ForwardDiff works as expected on CPUs, but yields a fairly cryptic error message on GPUs. Eventually I think that we will move to Enzyme, probably when the GPU support is ready. And thanks for working on that! It will truly bring this solver to a whole new feature level :)

File: src/util.jl (hunk 2 above)
Comment (b-fg) [reply src/util.jl:231]: Read your comment the other way around and didn't fix it in the end!

File: src/Flow.jl [hunk 3]
```diff
@@ -106,21 +108,19 @@ struct Flow{D, T, Sf<:AbstractArray{T}, Vf<:AbstractArray{T}, Tf<:AbstractArray{
...
         BC!(u,BCTuple(U,0.,D),exitBC,perdir); exitBC!(u,u,BCTuple(U,0.,D),0.)
         u⁰ = copy(u);
         fv, p, σ = zeros(T, Nd) |> f, zeros(T, Ng) |> f, zeros(T, Ng) |> f
         V, μ₀, μ₁ = zeros(T, Nd) |> f, ones(T, Nd) |> f, zeros(T, Ng..., D, D) |> f
+        BC!(μ₀,ntuple(zero, D),false,perdir)
         new{D,T,typeof(p),typeof(u),typeof(μ₁)}(u,u⁰,fv,p,σ,V,μ₀,μ₁,U,T[Δt],ν,g,exitBC,perdir)
     end
 end
 
-time(flow::Flow) = sum(flow.Δt[1:end-1])
```
Comment (b-fg) [inline src/Flow.jl:121]: @weymouth I was actually using this for time averaging stuff. Can we recover this and have `time(sim::Simulation)=time(sim.flow)` ?

File: src/util.jl (hunk 2 above)
Comment (b-fg) [reply src/util.jl:231]: Fixed in https://github.com/weymouth/WaterLily.jl/pull/131.

**Pull request merged**